serializer/src/main/java/org/apache/xml/serializer/Version.java

```diff
@@ -20,33 +20,97 @@
  */
 package org.apache.xml.serializer;
 
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
 /**
  * Administrative class to keep track of the version number of
  * the Serializer release.
  * <P>This class implements the upcoming standard of having
- * org.apache.project-name.Version.getVersion() be a standard way 
- * to get version information.</P> 
+ * org.apache.project-name.Version.getVersion() be a standard way
+ * to get version information.</P>
  * @xsl.usage general
  */
 public final class Version
 {
+  private static final String POM_PROPERTIES_JAR = "org/apache/xml/serializer/version.properties";
+  private static final String POM_PROPERTIES_FILE_SYSTEM = "serializer/target/classes" + POM_PROPERTIES_JAR;
+  private static final String VERSION_NUMBER_PATTERN = "^(\\d+)[.](\\d+)[.](D)?(\\d+)(-SNAPSHOT)?$";
+  private static final String NO_VERSION = "0.0.0";
+
+  private static String version = NO_VERSION;
+  private static int majorVersionNum;
+  private static int releaseVersionNum;
+  private static int maintenanceVersionNum;
+  private static int developmentVersionNum;
+
+  private static boolean snapshot;
```

Comment:
It looks like `serializer/Version` logic duplicates `xalan/Version` can it be shared?
`xalan` depends on `serializer` anyway, so it looks like `xalan` could just reuse the logic from `serializer` instead of duplicating it.

Reply:
No, it does not. I had checked that before.

```text
--- dependency:2.8:tree (default-cli) @ serializer ---        
xalan:serializer:jar:2.7.3                                    
+- org.apache.bcel:bcel:jar:6.7.0:compile                     
|  \- org.apache.commons:commons-lang3:jar:3.12.0:compile     
+- commons-logging:commons-logging:jar:1.2:provided           
+- javax:javaee-api:jar:6.0:provided                          
+- javax.servlet:servlet-api:jar:2.5:provided                 
+- xerces:xercesImpl:jar:2.12.2:provided                      
\- xml-apis:xml-apis:jar:1.4.01:provided                      
```

You see that I already optimised `XSLProcessorVersion` to avoid duplicate code. Again, I am pointing you to my PR description. Search for the paragraph starting with "There are 3 version classes". All those concerns you seem to have were already considered by me. Please do read descriptions too, not just review code in isolation.

Reply:
@kubycsolutions, @vlsi, do you think these version classes justify creating an extra module `xalan-commons` or so at this point, just to get rid of duplicate code for those non-essential classes? If you both say "yes", I shall of course follow your advice and add one such module. But it seems to be out of scope of this PR, which is meant to migrate an existing Ant solution to a Maven one with a different approach. OTOH, it is never too early to apply the DRY principle. Up to you, but you have to agree for me to change it.

Reply:
@kriegaex , sure xalan had no third-party dependencies before, so all these commons-logging and bcel is something unexpected. I assume they were introduced in the maven migration by mistake.
Here are the proper dependencies: https://repo1.maven.org/maven2/xalan/xalan/2.7.1/xalan-2.7.1.pom

Reply:
Frankly, I doubt that "version class" alone justifies introducing a `xalan-commons` module. `xalan:serializer` has always been a mandatory dependency, so placing `Version` class into `serializer` would be less invasive.

Reply:
Unfortunately, separate Version classes are necessary for the processor and Serializer. These are separate jarfiles because they are sometimes used independently. Hence we have to be able to ask each for it's version when the other is not available, which makes factoring out a common superclass across them undoable unless it's in a third jarfile. There's also the possibility of an out-of-sync pair, which we want to make visible. I can't see introducing a jarfile just for this, and I don't know of a Version in our dependencies which we could subclass productively. 

But if there's a separate Version for the xsltc compiler (I'm away from my desk and can't check easily, but I think there was/is),  _that_ can just proxy or subclass Xalan's Version, since they're the same jarfile and now report the same release version. That separation would exist only for historical reasons; the compiler started as a separate project before we adapted it to share Xalan's front end. I'd vote for deprecating this in favor of the Xalan Version, if it does exist. 

Of course the fine-tuning can be done as a separate PR, if folks are amenable to that. I've just been holding off merge because I haven't found cycles to review it myself (I'll try to do that today) and because there's still some productive discussion here. Alex, if you'd prefer further noncritical changes to be a separate PR,  we can open a Jira item for that and freeze this PR except for actual problems within its intended scope; I don't want to keep you hanging on this longer than necessary.

Reply:
@jkesselm , xalan.jar can't work without serializer.jar as it will almost definitely produce `NoClassDefFoundError`
Here is the relevant issue https://issues.apache.org/jira/browse/XALANJ-2649